### PR TITLE
Hide details of CNDB callback context during cn_open()

### PR DESCRIPTION
**Pull request by alexttx**
Signed-off-by: Alex Tomlinson <4172734+alexttx@users.noreply.github.com>

### CODE REVIEW

File: lib/cn/cn.c [hunk 1]
```diff
@@ -773,28 +773,63 @@ cn_maint_task(struct work_struct *work)
...
     struct kvset *kvset;
     merr_t err;
 
-    tree = ctx->cn->cn_tree;
+    tree = ctx->tree;
 
     node = map_lookup_ptr(ctx->nodemap, km->km_nodeid);
     if (!node) {
-        node = cn_node_alloc(tree, km->km_nodeid);
+        node = cn_node_alloc(ctx->tree, km->km_nodeid);
```
Comment (beckerg) [inline lib/cn/cn.c:832]: why not just "tree" ?

File: lib/cn/cn.c [hunk 2]
```diff
@@ -773,28 +773,63 @@ cn_maint_task(struct work_struct *work)
...
+ * Callback invoked by cndb_cn_instantiate() to place kvsets into tree nodes.
+ *
+ * This callback is invoked once for each kvset in a KVS.  Each callback
+ * contains a node ID, a kvset ID, and other metadata needed to open the
+ * on-media kvset.  It opens the kvsets and adds them to tree nodes, creating
+ * the tree nodes as needed.
  */
 static merr_t
-cn_kvset_cb(void *arg, struct kvset_meta *km, u64 kvsetid)
+cndb_cn_callback(void *arg, struct kvset_meta *km, u64 kvsetid)
```
Comment (beckerg) [inline lib/cn/cn.c:820]: Cool.  I prob would have called it cndb_cn_instantiate_cb() :-)

**Review by beckerg — APPROVED** [review]

File: lib/cn/cn.c (hunk 1 above)
Comment (alexttx) [reply lib/cn/cn.c:832]: I started to get rid of local var tree, but then backed out.  Will proceed with getting rid of it.

File: lib/cn/cn.c (hunk 2 above)
Comment (alexttx) [reply lib/cn/cn.c:820]: I had `cndb_cn_instantiate_callback` but that seemed too long.  Plus "instantiate callback" makes it sound like this function is instantiating a callback, as opposed to being a callback for that function.

File: lib/cn/cn.c [hunk 3]
```diff
@@ -930,27 +965,20 @@ cn_open(
...
     }
 
-    ctx.cn = cn;
-    ctx.nodemap = map_create(CN_FANOUT_MAX);
-    if (ev(!ctx.nodemap)) {
-        err = merr(ENOMEM);
+    /* Build tree from data in CNDB.
+     */
+    err = cndb_cn_ctx_init(&ctx, cn->cn_tree, cn->cn_tree->ct_root);
+    if (ev(err))
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/cn/cn.c:971]: To insert the root node into the map.  A tiny bit cleaner than this code knowing how to find root in the tree structure.  Some unit test somewhere might thank me later.

File: lib/cn/cn.c [hunk 4]
```diff
@@ -773,28 +773,63 @@ cn_maint_task(struct work_struct *work)
...
+
+    nodemap = map_create(CN_FANOUT_MAX);
+    if (!nodemap)
+        return merr(ENOMEM);
+
+    ctx->nodemap = nodemap;
+    ctx->tree = tree;
+
+    /* Pre-populate node map with root node, which always has ID 0. */
+    map_insert_ptr(ctx->nodemap, 0, root);
```
Comment (nabeelmmd) [inline lib/cn/cn.c:797]: map_insert_ptr() can fail?

File: lib/cn/cn.c (hunk 2 above)
Comment (beckerg) [reply lib/cn/cn.c:820]: That's why I use _cb() :-)

File: lib/cn/cn.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/cn.c:971]: well, now you need an assert to check that they match...

[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/cn.c:971]: I hate our unit tests...

File: lib/cn/cn.c [hunk 5]
```diff
@@ -930,27 +965,20 @@ cn_open(
...
+
+    cndb_cn_ctx_fini(&ctx);
 
-    err = cndb_cn_instantiate(cndb, cnid, &ctx, cn_kvset_cb);
     if (ev(err))
         goto err_exit;
 
-    /* Walk the list of nodes created/populated by cn_kvset_cb() and insert
-     * all leaf nodes into the route map (i.e., all nodes except the root
-     * node which always has node ID 0).
```
Comment (beckerg) [inline lib/cn/cn.c:953]: I think this is an pretty informative comment for someone wondering how the list got built...

**Review by nabeelmmd — APPROVED** [review]

File: lib/cn/cn.c (hunk 1 above)
Comment (beckerg) [reply lib/cn/cn.c:832]: We could probably have just added a function to the route map to find an entry by node ID, then we could just get rid of all this map stuff...

File: lib/cn/cn.c (hunk 4 above)
Comment (beckerg) [reply lib/cn/cn.c:797]: I'm liking the idea more and more to have a route map iterator that calls the caller's comparator to find nodes of interest...

File: lib/cn/cn.c (hunk 1 above)
Comment (alexttx) [reply lib/cn/cn.c:831]: This way we build nodes, then create route map / tree.  Which is kind of nice.  Baby stepping our way through cn_open.

File: lib/cn/cn.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/cn/cn.c:970]: You mean that ct_root matches passed in root?  Nah.  If caller does that they have other issues.

File: lib/cn/cn.c (hunk 4 above)
Comment (alexttx) [reply lib/cn/cn.c:799]: Memory allocation can cause failure.

Comment (nabeelmmd) [reply lib/cn/cn.c:799]: return map_insert_ptr(ctx->nodemap, 0, root);

File: lib/cn/cn.c (hunk 5 above)
Comment (alexttx) [reply lib/cn/cn.c:953]: Will revert.

**Comment by alexttx** [comment]
I removed `ct_dgen_init` member from `struct cn_tree` and track it directly in the CNDB callback context.  Then use that to set the `cn_ingest_dgen`.

File: lib/cn/cn.c (hunk 1 above)
Comment (beckerg) [reply lib/cn/cn.c:836]: Yeah, we had a bug where I didn't wait for all the kvsets to be added before looking for the max key, so we kinda have to do it this way.  I suppose you could update the max key on each kvset but that'd be silly..

**Pull request merged**